### ISSUE TITLE
kernighan lin optimization

### ISSUE THREAD

**Issue by zcollins0** [networkx/networkx#3330]
I have noticed that the NetworkX implementation of Kernighan-Lin is missing the optimizations suggested by here: https://dl.acm.org/citation.cfm?id=259855 They are proven to produce exactly the same result, and contribute a massive speedup to KL. In particular, the column scanning method would be very easy to implement. The NetworkX KL implementation is too slow to use for any sizable graph, and the column scanning optimization should increase its usability to graphs of 10x the current size.

I intend to make a pull request adding column scanning to KL. Would this be more appropriate as an additional algorithm, or is it ok as a change to KL?

**Comment by dschult** [comment]
This sounds great -- As you dive into creating the new data structure for Quick Cut, I would suggest that you try to build it out of native Python structures (like dict) wherever possible. They are quite fast. As for how to name them, why don't you (for now) rename the existing as ```original_ker...``` and use the new methods with the current name. That way we can test against the older method easily and users have both available if they need/want to compare.

**Comment by zcollins0** [comment]
Sure! I'm prepping for my thesis defense the next few weeks, so I won't have a whole lot of time. Hopefully I can have a good deal of this done by the end of March.

**Comment by boothby** [comment]
@zcollins0 are you still planning to work on this?  I recently became annoyed with how slow the K-L implementation is, and I can take this over if you're busy.

**Comment by boothby** [comment]
I implemented the column and diagonal scanning algorithms, as well as an even simpler sweep which alternates between sides to move single nodes to/from.  This "best node first" strategy appears to be common practice.  I tried a bunch of examples, dense and sparse, and didn't observe a significant difference in the distribution of cuts produced by these methods.  The best-first algorithm has the simplest code by far, and ran the fastest overall.

As per @dschult 's comment, I used pre-existing data structures.  I did limited testing of both the PairingHeap and BinaryHeap (with a handful of dense and sparse graphs), and the BinaryHeap won out by about 10%.
